### ISSUE TITLE
Mypy incorrectly passes the type check `Iterable[T]` with type `numpy.NDArray[Any]`

### ISSUE THREAD

**Issue by gnaggnoyil** [python/mypy#15033]
**Bug Report**

When calling a function requiring its parameter be `Iterable[T]` with an argument of type `numpy.NDArray[Any]` or `numpy.NDArray[U]`, mypy falsely passes the check and reports success for that.

**To Reproduce**

```python
import typing

import numpy


def foo(x: typing.Iterable[int]):
    for xv in x:
        if not isinstance(xv, int):
            raise TypeError("")

foo(numpy.zeros((3, ), dtype=float))
```

**Expected Behavior**

Mypy should output errors like `error: Argument 1 to "foo" has incompatible type "ndarray[Any, dtype[Any]]"; expected "Iterable[int]"`

**Actual Behavior**

Mypy passes the check without errors and reported `Success: no issues found in 1 source file`.

**Your Environment**

- Mypy version used: `mypy 1.1.1 (compiled: no)`
- Python version used: `Python 3.10.10`
- Numpy version used: 1.24.2, with `numpy.typing` installed.

(I ran all the above steps in Archlinux environment and all relevant packages are installed through Archlinux official repository)

**Comment by tmke8** [comment]
I don't think that's mypy's fault. The type stubs for numpy define `__iter__` on `ndarray` here: https://github.com/numpy/numpy/blob/26de8cb3c19361c45df0e8ca43945a687e213d43/numpy/__init__.pyi#L1837

I think it should actually be
```python
    def __iter__(self) -> Iterable[_DType_co]: ...
```

Then the `dtype` would be used for checking the compatibility with `Iterable[int]`.

**Comment by JelleZijlstra** [comment]
Numpy arrays are indeed iterables. Possibly we should infer an array of floats instead of ints, but I would guess that is something to fix in numpy's stubs, not a bug in mypy.
